pkg/controller/servingcert/controller/stateful_secret_creating_controller.go

```diff
@@ -0,0 +1,217 @@
+	}
+
+	if !sc.requiresCertGeneration(sharedStatefulSet) {
+		return nil
+	}
+
+	// make a copy to avoid mutating cache state
+	statefulSetCopy := sharedStatefulSet.DeepCopy()
+	return sc.generateCert(ctx, statefulSetCopy)
+}
+
+func (sc *statefulSetServingCertController) generateCert(ctx context.Context, statefulSetCopy *appsv1.StatefulSet) error {
+	klog.V(4).Infof("generating new cert for %s/%s", statefulSetCopy.GetNamespace(), statefulSetCopy.GetName())
+	if statefulSetCopy.Annotations == nil {
+		statefulSetCopy.Annotations = map[string]string{}
+	}
+
+	secret := statefulSetToBaseSecret(statefulSetCopy)
+	if err := regenerateStatefulSetSecret(sc.servingCA, statefulSetCopy, secret); err != nil {
+		return err
+	}
+
+	_, err := sc.secretClient.Secrets(statefulSetCopy.Namespace).Create(ctx, secret, metav1.CreateOptions{})
+	if err != nil && !kapierrors.IsAlreadyExists(err) {
+		return sc.updateStatefulSetFailure(ctx, statefulSetCopy, err)
+	}
+	if kapierrors.IsAlreadyExists(err) {
+		actualSecret, err := sc.secretClient.Secrets(statefulSetCopy.Namespace).Get(ctx, secret.Name, metav1.GetOptions{})
+		if err != nil {
+			return sc.updateStatefulSetFailure(ctx, statefulSetCopy, err)
+		}
+
+		if actualSecret.Annotations[api.StatefulSetUIDAnnotation] != string(statefulSetCopy.UID) {
+			uidErr := fmt.Errorf("secret %s/%s does not have corresponding StatefulSet UID %v", actualSecret.GetNamespace(), actualSecret.GetName(), statefulSetCopy.UID)
+			return sc.updateStatefulSetFailure(ctx, statefulSetCopy, uidErr)
+		}
+		klog.V(4).Infof("renewing cert in existing secret %s/%s", secret.GetNamespace(), secret.GetName())
+		// Actually update the secret in the regeneration case (the secret already exists but we want to update to a new cert).
+		_, updateErr := sc.secretClient.Secrets(secret.GetNamespace()).Update(ctx, secret, metav1.UpdateOptions{})
+		if updateErr != nil {
+			return sc.updateStatefulSetFailure(ctx, statefulSetCopy, updateErr)
+		}
+	}
+
+	sc.resetStatefulSetAnnotations(statefulSetCopy)
+	_, err = sc.statefulSetClient.StatefulSets(statefulSetCopy.Namespace).Update(ctx, statefulSetCopy, metav1.UpdateOptions{})
+
+	return err
+}
+
+func getStatefulSetNumFailures(statefulSet *appsv1.StatefulSet) int {
+	numFailuresString := statefulSet.Annotations[api.ServingCertErrorNumAnnotation]
+	if len(numFailuresString) == 0 {
+		return 0
+	}
+
+	numFailures, err := strconv.Atoi(numFailuresString)
+	if err != nil {
+		return 0
+	}
+
+	return numFailures
+}
+
+func (sc *statefulSetServingCertController) requiresCertGeneration(statefulSet *appsv1.StatefulSet) bool {
+	// check the secret since it could not have been created yet
+	secretName := statefulSet.Annotations[api.ServingCertSecretAnnotation]
+	if len(secretName) == 0 {
+		return false
+	}
+
+	secret, err := sc.secretLister.Secrets(statefulSet.Namespace).Get(secretName)
+	if kapierrors.IsNotFound(err) {
+		// we have not created the secret yet
+		return true
```

Comment:
Note to self: This ignores `getStatefulSetNumFailures` and will keep trying to create the secret (e.g. if it is too large).

Reply:
Fixed.